### PR TITLE
Always create the main artifact along with a shim-classifier artifact [databricks]

### LINKED ISSUE #8156: [BUG] Install phase for modules with Spark build classifier fails for install plugin versions 3.0.0+
**Describe the bug**

We do not specify the maven-install-plugin version in the parent pom. Thus Maven uses the default version from https://github.com/apache/maven/blob/cecedd343002696d0abb50b32b541b8a6ba2883f/maven-core/src/main/resources/META-INF/plexus/default-bindings.xml#L44

If the user ends using Maven with 2.4 install plugin we get an info message

```
[INFO] No primary artifact to install, installing attached artifacts instead.
```

and the build proceeds.

However, more recent Maven versions e.g. 3.9.1 use the 3.1.0 install plugn. And the build fails with

```
 The packaging plugin for this project did not assign a main file to the project but it has attachments.
```

So we should should do one of the following:

* always produce a dummy primary no-classifier artifact for submodules
* execute the install phase using the `install-file` goal as we do in the dist submodule which is [documented for the secondary artifacts](https://maven.apache.org/plugins/maven-install-plugin/examples/installing-secondary-artifacts.html)
* pin the version of the install plugin to the most recent version not throwing the error

**Steps/Code to reproduce bug**

Run `mvn install` using a Maven 3.9.1 or simply declare install plugin version 3.1.0 in the parent `./pom.xml`

**Expected behavior**

The build should work for users with recent Maven versions.

**Environment details (please complete the following information)**
- local dev machine

**Additional context**

Failure originally reported @NVnavkumar

### PR DESCRIPTION
**Pull request by gerashegalov**
Fixes #8156 

Make sure that the main artifact is always produced even when the actual output  is the shim artifact
 
- Set install plugin version to 3.1.1
- Set jar plugin version to 3.3.0
- Create a generic maven-jar-plugin definition
- Configure modules that need a shim artifact via params
- Create an install check with recent patches to major Maven versions
- Add enforcer plugin, currently just validating the minimum version for Maven

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

File: pom.xml [hunk 1]
```diff
@@ -679,8 +679,8 @@
         <parquet.hadoop.version>1.10.1</parquet.hadoop.version>
         <spark.version.classifier>spark${buildver}</spark.version.classifier>
         <cuda.version>cuda11</cuda.version>
-        <spark-rapids-jni.version>23.08.0-SNAPSHOT</spark-rapids-jni.version>
-        <spark-rapids-private.version>23.08.0-SNAPSHOT</spark-rapids-private.version>
+        <spark-rapids-jni.version>${project.version}</spark-rapids-jni.version>
+        <spark-rapids-private.version>${project.version}</spark-rapids-private.version>
```
Comment (jlowe) [inline pom.xml:683]: This change is not desired.  We change these versions separate from the project version (e.g.: these dependencies stop being a snapshot before this project does).

File: dist/pom.xml [hunk 2]
```diff
@@ -577,7 +577,6 @@ self.log("... OK")
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-install-plugin</artifactId>
-                <version>3.0.1</version>
```
Comment (jlowe) [inline dist/pom.xml:580]: Any idea why this version was originally requested?  Curious if there was a historic reason to pick this specific version here and if we're going to expose a known bug by rolling back.

**Review by jlowe — COMMENTED** [review]
premerge build has been stuck for many hours, appears unable to launch containers due to insufficient resources. cc: @pxLi @NvTimLiu

File: dist/pom.xml (hunk 2 above)
Comment (gerashegalov) [reply dist/pom.xml:580]: Ouch, thanks for catching. Pretty sure I needed this for pomFile read from the jar.

**Comment by pxLi** [comment]
> 

checked seems intermittent blossom scheduler issue. We will keep monitoring

**Comment by pxLi** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Review by jlowe — APPROVED** [review]
Changes look good but CI is blocked by #8702

**Pull request merged**